### ISSUE TITLE
compile fails randomly with "TypeError: can't convert complex to float"

### ISSUE THREAD

**Issue by nelimeee** [QISKit/qiskit-terra#593]
### Informations

- **QISKit (Python SDK) version**: qiskit==0.5.4
- **Python version**: Python 3.6.5
- **Operating system**: Fedora 26

### What is the current behavior?

The `compile` method randomly fails to compile and throw a `TypeError` exception.
Here is the full stack-trace obtained from the [QISKit_issue_593.py](https://gist.github.com/nelimeee/0bc39528bfa2c5da4692785e1d3ec153) file:

```
$ python3 compiler_error.py 
Compilation number 0...
The error happened!
Traceback (most recent call last):
  File "compiler_error.py", line 27, in <module>
    raise e
  File "compiler_error.py", line 24, in <module>
    qiskit.compile(unroller.backend.circuit, ibmqx5)
  File "/[...]/qiskit/wrapper/_wrapper.py", line 170, in compile
    skip_transpiler)
  File "/[...]qiskit/_compiler.py", line 154, in compile
    json_circuit = DagUnroller(dag_circuit, JsonBackend(dag_circuit.basis)).execute()
  File "/[...]/qiskit/unroll/_dagunroller.py", line 38, in execute
    self._process()
  File "/[...]/qiskit/unroll/_dagunroller.py", line 185, in _process
    current_node["qargs"])
  File "/[...]/qiskit/unroll/_jsonbackend.py", line 277, in start_gate
    args)),
  File "/[...]/qiskit/unroll/_jsonbackend.py", line 276, in <lambda>
    'params': list(map(lambda x: float(x.real(nested_scope)),
  File "/[...]/qiskit/qasm/_node/_real.py", line 53, in real
    return float(self.value.evalf())
  File "/[...]/sympy/core/expr.py", line 238, in __float__
    raise TypeError("can't convert complex to float")
TypeError: can't convert complex to float
```

After some investigations, the "complex" value is close to `4.71 - 2.1e-8`. Because the compiler is not deterministic, results are not really reproducible and so you may obtain a different value.

### Steps to reproduce the problem

1. Download the [QISKit_issue_593.py](https://gist.github.com/nelimeee/0bc39528bfa2c5da4692785e1d3ec153) file.
2. Download the [4x4.qasm](https://gist.github.com/nelimeee/31a84435ad649471405d2afc3d9b13ff) file and save it in a file named `4x4.qasm`, in the same folder as the script.
3. In your qiskit installation, **temporarily** change the default value of `eps` in the `qiskit.mapper._mapping.yzy_to_zyz` function to `2e-8`.
 **Don't forget to write back the initial default value which is `1e-9` after testing.**
4. `python3 QISKit_issue_593.py`

The script may take several minutes to finish: one compilation is approximately 20 seconds on my computer and the exception may occur only after 4 or 5 steps.

### What is the expected behavior?

With the modification of `eps` in `qiskit.mapper._mapping.yzy_to_zyz`, this script should run for ever without being interrupted by an exception thrown by the compiler.

### Suggested solutions

To recall, the "complex" value responsible for the exception is close to `4.71 - 2.1e-8`. With such a tiny amplitude, I think the imaginary part can be the result of a mathematical operation with bad rounding. @rraymondhp proposed the computation of eigenvalues on Slack.

EDIT: Updated to link to gists instead of full code in the description.

**Comment by ajavadia** [comment]
Possibly the same issue as #239.

**Comment by jaygambetta** [comment]
This is the same issue as what #607 will fix.

**Comment by jaygambetta** [comment]
@ajavadia can we check if this can be closed now.

**Comment by nonhermitian** [comment]
Should be.  Same as #520.  Becuase rotations not valid, so small complex term.

**Comment by ajavadia** [comment]
Yes, I just confirmed that this is also fixed by #626.
